frame/elections-phragmen/src/lib.rs

```diff
@@ -656,93 +751,135 @@ decl_module! {
-		/// `NewTerm(\[\])`. See the description of `NewTerm`.
-		EmptyTerm,
-		/// Internal error happened while trying to perform election.
-		ElectionError,
-		/// A \[member\] has been removed. This should always be followed by either `NewTerm` or
-		/// `EmptyTerm`.
-		MemberKicked(AccountId),
-		/// A candidate was slashed due to failing to obtain a seat as member or runner-up
-		CandidateSlashed(AccountId, Balance),
-		/// A seat holder (member or runner-up) was slashed due to failing to retaining their position.
-		SeatHolderSlashed(AccountId, Balance),
-		/// A \[member\] has renounced their candidacy.
-		MemberRenounced(AccountId),
-		/// A voter was reported with the the report being successful or not.
-		/// \[voter, reporter, success\]
-		VoterReported(AccountId, AccountId, bool),
+impl<T: Trait> Module<T> {
+	/// The deposit value of `count` votes.
+	fn deposit_of(count: usize) -> BalanceOf<T> {
+		T::VotingBondBase::get().saturating_add(
+			T::VotingBondFactor::get().saturating_mul((count as u32).into())
+		)
 	}
-);
 
-impl<T: Trait> Module<T> {
-	/// Attempts to remove a member `who`. If a runner-up exists, it is used as the replacement and
-	/// Ok(true). is returned.
+	/// Attempts to remove a member `who`. If a runner-up exists, it is used as the replacement.
 	///
-	/// Otherwise, `Ok(false)` is returned to signal the caller.
+	/// Returns:
 	///
-	/// If a replacement exists, `Members` and `RunnersUp` storage is updated, where the first
-	/// element of `RunnersUp` is used as the replacement and `Ok(true)` is returned. Else,
-	/// `Ok(false)` is returned with no storage updated.
+	/// - `Ok(true)` if the member was removed and a replacement was found.
+	/// - `Ok(false)` if the member was removed and but no replacement was found.
+	/// - `Err(_)` if the member was no found.
 	///
-	/// Note that this function _will_ call into `T::ChangeMembers` in case any change happens
-	/// (`Ok(true)`).
+	/// Both `Members` and `RunnersUp` storage is updated accordingly. `T::ChangeMember` is called
+	/// if needed. If `slash` is true, the deposit of the potentially removed member is slashed,
+	/// else it is unreserved.
 	///
-	/// If replacement exists, this will read and write from/into both `Members` and `RunnersUp`.
-	fn remove_and_replace_member(who: &T::AccountId) -> Result<bool, DispatchError> {
-		let mut members_with_stake = Self::members();
-		if let Ok(index) = members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(who)) {
-			members_with_stake.remove(index);
-
-			let next_up = <RunnersUp<T>>::mutate(|runners_up| runners_up.pop());
-			let maybe_replacement = next_up.and_then(|(replacement, stake)|
-				members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(&replacement))
+	/// ### Note
+	///
+	/// This function attempts to preserve the prime. If the removed members is not the prime, it is
+	/// set again via [`Trait::ChangeMembers` ] to prevent it being wiped.
```

Comment:
If I understand correclty the prime is reset when Members are changed in order to preserve the prime in case the implementation of ChangeMembers trait only refer to the prime as an index inside the member set ?

I worry this specific usage of the trait can be easily missed as it isn't written anywhere else as far as I see.
This requirements could be written either:
* in the doc of `ChangeMembers`
* or in the associated type ChangeMembers saying that on top of ChangeMembers assumption the trait is used with this constraint.

Reply:
> in case the implementation of ChangeMembers trait only refer to the prime as an index inside the member set?

Hmm I didn't think about this at all.. yeah this should be documented. 

But still my usage should be safe, I assume? I am first setting the new members and then set the prime again. Even if an index has changed, then the new prime should be correct. 

I recall that we had such problems in kusama: if a members renounces, then we lose the prime until the next election. This kinda sucks.

Reply:
yes implementation is fine, I just worry those assumption gets expected somewhere else and then broke because ppl missed them.

Reply:
I'm finally ok, I feel this function is slightly overdoing, but that shouldn't lead ppl to think this property is ensured by election-phragmen usages nor ChangeMembers.

Reply:
ah I finally understand where i'm wrong. change_member_sorted reset the prime, thus it needs to be set again. OK all good then